### PR TITLE
Raise informative error when a converted checkpoint matches none of the loaded class's weights

### LINKED ISSUE #47405: Silent conversion skip on model_type mismatch: loading a converted checkpoint via the wrong Auto class yields randomly-initialized weights that generate fluently
### System Info

- `transformers` version: 5.14.1
- Platform: Linux-6.17.0-1007-aws-x86_64-with-glibc2.39
- Python version: 3.12.3
- Huggingface_hub version: 1.24.0
- Safetensors version: 0.8.0
- Accelerate version: 1.14.0
- Accelerate config: not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.12.0a0+git220ecc2 (CUDA)
- Using distributed or parallel set-up in script?: No (single process; reproduces with device_map="auto" or single GPU)
- Using GPU in script?: Yes
- GPU type: NVIDIA B300 SXM6 AC
Supplementary: transformers 5.14.1; torch 2.12.0a0 cu12.9; model: Inkling-975B (multimodal, `model_type: inkling_mm_model`, converted checkpoint).

### Who can help?

@CyrilVallez (model loading / conversion mapping) @zucchini-nlp (multimodal)
 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Checkpoint conversion mappings are resolved by the **loaded class's** expectations; when a user loads a converted checkpoint through a class whose naming doesn't match (e.g. `AutoModelForCausalLM` on a checkpoint whose weights live under the multimodal layout), no conversion table applies and every checkpoint tensor fails to match:
 
```python
from transformers import AutoModelForCausalLM
m = AutoModelForCausalLM.from_pretrained(PATH_TO_INKLING, dtype="bfloat16")  # multimodal ckpt
out = m.generate(**tok("The capital of France is", return_tensors="pt"), max_new_tokens=8)
# -> fluent but random text ("empatan ..."); no error raised
```
Symptoms in the log: "load" completes in seconds for a 1.9 TB checkpoint; conversion progress shows `0it`; the only signals are symmetric MISSING/UNEXPECTED key warnings (easily lost in noise) — then the model **generates fluent garbage** from random init. Cost of the failure mode: it looks like a working model with bad weights, sending users into weight-level debugging instead of a one-line class fix.

### Expected behavior

Hard error (or at minimum a prominent warning) when: the checkpoint's `config.model_type` has a registered conversion mapping, AND the resolved class matches ~0% of checkpoint tensors. "You loaded a `<model_type>` checkpoint via `<class>`; its weights require the `<other class>` conversion mapping" would turn hours of debugging into seconds. Secondary question: is the absence of a text-only `AutoModelForCausalLM` load path for this checkpoint layout an intentional policy or an omission?

### PR DESCRIPTION
**Pull request by SelfSL**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47408)

Fixes #47405

**Failure mode:** loading a checkpoint whose architecture has registered weight conversions through a non-matching class (e.g. `AutoModelForCausalLM` on a multimodal layout) currently succeeds silently: zero tensors match, the model is fully randomly initialized, and it *generates fluently* -- sending users into weight-level debugging instead of a one-line class fix.

**Change:** in `_finalize_model_loading`, after `_adjust_missing_and_unexpected_keys`, raise a `ValueError` when weight conversions are registered (`load_config.weight_mapping`), there are unexpected checkpoint keys, and >=99% of the model's keys are missing. The error names both counts and points at the Auto-class fix.

Design notes:
- **Outcome-based predicate** (near-total mismatch + conversions present) rather than probing the checkpoint's `model_type` registry -- covers every route into the failure, and cannot fire on any sane partial load.
- **Raise inside the `try`** so the `finally` still emits the full loading report before the exception -- verified: the MISSING/UNEXPECTED table prints, then the error surfaces.
- **No bypass kwarg**: a ~100%-miss load has no legitimate use, and `ignore_mismatched_sizes` has different semantics (shape mismatches). Happy to add an escape hatch if preferred.

Tests: positive (raises with conversions + total mismatch) and negative control (identical mismatch without conversions does not raise) in `tests/utils/test_modeling_utils.py`.

cc @CyrilVallez @zucchini-nlp

---
_Local `make check-repo` note: the `types` check flags a pre-existing `torch.cat(..., axis=-1)` at `modeling_utils.py:985` on clean main; unrelated to this diff. All ruff/style/consistency checks pass._

### CODE REVIEW

**Review by webtecnica — COMMENTED** [review]
Great fix for a nasty silent-init bug. I've reviewed the diff and verified the test indentation via the raw source — the alignment is correct.

**Design:**
- Outcome-based predicate (near-total mismatch + conversions) is the right approach — much more robust than probing model_type.
- The placement inside the `try` block is correct: `finally` still emits the full loading report before the error surfaces.
- The 99% threshold is safe: no legitimate partial load loses 99%+ of weights.

**Suggestions for consideration:**

1. 🗝️ **Escape hatch** — Consider adding `_force_load=True` (private, underscore-prefixed) as an opt-out. Even if undocumented, it keeps flexibility for edge cases without exposing a public API.

2. 🔌 **Quantizer coverage** — The same silent-init pattern can occur with `CompressedTensorsConfig` (#47407) where `hf_quantizer` is set but `weight_mapping` may be empty. Extending to also check `load_config.hf_quantizer` would prevent that class of failure too.

3. ✅ **Tests** — Positive and negative controls are well-structured. The unit-level approach (tiny model + synthetic LoadStateDictInfo) is clean and avoids external dependencies.

Happy to help with any follow-up on these points.

**Comment by zucchini-nlp** [comment]
Not really what we want and the error exists in old released versions only!
